### ISSUE TITLE
[feature] Add option to choose location of where to save the file

### ISSUE THREAD

**Issue by sga-13** [gabm/Satty#92]
The title. Can an option (either graphical or keyboard shortcut like Ctrl+Shift+S or something alike) which will open a file picker (maybe the gtk one or the xdg file picker) and we are able to choose a location to save the output, which may or may not be independent of where the Ctrl+S would save

I tried to achieve something similar by using yazi (a terminal file manager) to act as a file picker, and using that either choose the location to save or the exact file path, but somehow i am not able to stop it from rendering before the screenshot is taken

I have tried to 

```
grim -g "$(slurp)" - | satty --filename - --annotation-size-factor 0.4 --output-filename "$(yazi-filepicker)/out.png"
```

and other variations where i give full path, or add a sleep time (which works but limits me on how fast should i snip)

any help is welcomed, and if this issue is duplicate, then I am sorry, but I could not find any

**Comment by gabm** [comment]
This feature is not present (yet). Satty started life as a CLI tool without graphical interaction apart from the annotation task itself. But it developed since and the lack of a graphical file picker for saving is something that is lacking - i fully agree.

To preserve the current functionality, I suggest the following mechanism:

- if the user specified a save path or pattern, then use that one as "is"
- if the user did not specify an output filename, then we show the button anyway (contrary to now) and upon button press, show the gtk file picker and then save the file there...

I think that would be quite easy to implement - do you want to take a shot? I can provide guidance

**Comment by sga-13** [comment]
I have no experience in rust, all i can do is python and bash scripting (and weirdly java), I can try to take a shot at it, but please do not expect much

i am going to imagine that we include some function here (like color palette has something set by default, but instead calling in filepicker)

```
impl Default for Configuration {
    fn default() -> Self {
        Self {
            input_filename: String::new(),
            output_filename: None,
            fullscreen: false,
            early_exit: false,
            initial_tool: Tools::Pointer,
            copy_command: None,
            annotation_size_factor: 1.0,
            save_after_copy: false,
            color_palette: ColorPalette::default(),
            default_hide_toolbars: false,
            font: FontConfiguration::default(),
        }
    }
}
```

if this is correct, then I can try to take a shot, if not please guide me

**Comment by nomm-dot** [comment]
Just wanted add more interest in this functionality, as satty is great for taking quick screenshots while taking notes. I like to use emacs org, and have an /images dir per project/course/topic to reference the image for more mobility of the notes. Additionally, with the functionality to change the name when choosing the directory would be great. This is mostly so that I can add a keyword to the file name so that I don't need to use a GUI to know what screenshot I'm moving.

**Comment by gabm** [comment]
I do see the need for that and I'll happily accept PRs along the mentioned lines!

**Comment by RobertMueller2** [comment]
I also think we should add a file chooser at some point, but this can be scripted right now. (Perhaps could use action-on-enter or similar.) Pipe or substitution does not work, but a shell script can run yazi after satty.

```
#!/bin/sh

tmp1="$(mktemp).png"
satty --filename - --output-filename "${tmp1}"

tmp2=$(mktemp)
yazi --chooser-file="$tmp2" .

filename=$(cat "$tmp2")
rm "$tmp2"

[ -z "$filename" ] && {
  echo "No file selected."
  rm "$tmp1"
  exit 1
}

mv "$tmp1" "$filename"
```

e.g.
`grim -g "$(slurp)" - |  ~/tmp/testscript.sh`

**Comment by Xismy** [comment]
#246 seems to be about a similar feature. In that PR, @gabm suggested using the gtk4 FileDialog. Since the author mentioned having dropped development, I would like to take it on.

I have a working [branch in my fork](https://github.com/Xismy/Satty/tree/feature/save-as) that adds a "Save As" button using FileDialog, as suggested. However, this struct requires gtk 4.10, which deprecates some functions, filling the build output with warnings.

I can make a new PR with this, or try with FileChooserDialog, which was supported prior to 4.10 (and is actually deprecated in 4.10).

**Comment by RobertMueller2** [comment]
> [#246](https://github.com/gabm/Satty/pull/246) seems to be about a similar feature. In that PR, [@gabm](https://github.com/gabm) suggested using the gtk4 FileDialog. Since the author mentioned having dropped development, I would like to take it on.
> 
> I have a working [branch in my fork](https://github.com/Xismy/Satty/tree/feature/save-as) that adds a "Save As" button using FileDialog, as suggested. However, this struct requires gtk 4.10, which deprecates some functions, filling the build output with warnings.
> 
> I can make a new PR with this, or try with FileChooserDialog, which was supported prior to 4.10 (and is actually deprecated in 4.10).

Yeah, I stumbled across the FileChooserDialog and FileDialog before and wondered the same. IMO unless Relm4 changes anything, we shouldn't use a different GTK feature set, so I'd go with FileChooserDialog for now even though we know we'll have to replace it eventually.

**Comment by RobertMueller2** [comment]
Added with #292, so I'll close this.
